wasm_proto/core/util.c

```diff
@@ -0,0 +1,27 @@
+/*
+ * `util.c` defines some utility functions
+ */
+
+#include "core.h"
+#include "util.h"
+#include "uint.h"
+
+void setcc(uint rslt) {
+	uint cc = 0x2;
+	if(rslt > 0) {
+		cc >>= 1;
+	} else if(rslt < 0) {
+		cc <<= 1;
+	}
+
+	core.psr = core.psr & ~0x7 | cc;
+}
+
+uint sext16(uint val, uint w) {
+	return ~(~0<<w) & val | (1<<(w-1) & val ? ~(~(uint)0>>w) : 0);
+}
+
+uint trunc16(uint val) {
```

Comment:
do we need this function? u can leave it regardless.

kinda related: should we use `uint8_t` instead of `uint`? I'm a little rusty on C, and maybe webasm wouldn't play well with it? if u don't have a quick answer, I can look into this more.

Reply:
Every source I find seems to say that the Emscripten compiler can handle the entire C standard library, so using a stdint type should be fine. That being said, the Wasm website says:

> WebAssembly has 32-bit and 64-bit architecture variants, called wasm32 and wasm64. wasm32 has an ILP32 data model, meaning that int, long, and pointer types are all 32-bit, while the long long type is 64-bit. wasm64 has an LP64 data model, meaning that long and pointer types will be 64-bit, while int is 32-bit.

Right now, only wasm32 is available. I chose to use `int` and `uint` (i.e. `unsigned int`) because it had a direct mapping to a Wasm type while something like `uint16_t` wouldn't. If `emcc` claims to be able to handle the entire standard library, then it must be able to use `uint16_t` or something of the like, though, so I'll change it over now.

Reply:
As for whether we need the `trunc` function, if we use a 32-bit type, bits left in the upper 16 bits from previously overflowing calculations (that is, overflowing the bottom 16 bits) can cause overflow later when the lower 16 bits otherwise shouldn't. If we use `uint16_t`, we wouldn't need it though.